### PR TITLE
authentication_with_omniauth

### PR DESCRIPTION
**Pull request by sujoydatta**

### CODE REVIEW

File: app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def all
+    user = User.from_omniauth(request.env["Omniauth.auth"])
```
Comment (SSalekin) [inline app/controllers/omniauth_callbacks_controller.rb]: remove ()

File: app/views/devise/registrations/edit.html.erb [hunk 2]
```diff
@@ -12,20 +12,22 @@
...
-    <%= f.password_field :password_confirmation, autocomplete: "off" %>
-  </div>
-
-  <div class="field">
-    <%= f.label :current_password %> <br />
-    <%= f.password_field :current_password, autocomplete: "off" %>
-  </div>
+  <% if f.object.update_with_password? %>
+    <div class="field">
+      <%= f.label :password %> <i>(leave blank if you don't want to change it)</i><br />
```
Comment (MuhammadTamzid) [inline app/views/devise/registrations/edit.html.erb]: i18n

File: app/models/user.rb [hunk 3]
```diff
@@ -1,10 +1,43 @@
...
   after_initialize :init_admin
 
-  devise :database_authenticatable, :registerable,
+  devise :database_authenticatable, :registerable, :omniauthable,
          :recoverable, :rememberable, :trackable, :validatable
   has_many :lessons
 
+  class << self
+    def from_omniauth auth
+      where(auth.slice(:provider, :uid)).first_or_create do |user|
```
Comment (SSalekin) [inline app/models/user.rb]: remove ()

Comment (SSalekin) [inline app/models/user.rb]: remove ()

File: app/views/devise/registrations/new.html.erb [hunk 4]
```diff
@@ -12,19 +12,20 @@
     <%= f.label :email %><br />
     <%= f.email_field :email, autofocus: true, class: "form-control" %>
   </div>
+  <% if f.object.password_required? %>
+    <div class="field">
+      <%= f.label :password %>
+      <% if @minimum_password_length %>
+      <em>(<%= @minimum_password_length %> characters minimum)</em>
```
Comment (MuhammadTamzid) [inline app/views/devise/registrations/new.html.erb]: i18n

**Comment by MuhammadTamzid** [comment]
Ok

**Comment by SSalekin** [comment]
Ok

**Comment by MuhammadTamzid** [comment]
Ok

File: config/initializers/devise.rb [hunk 5]
```diff
@@ -236,6 +236,12 @@
   # Add a new OmniAuth provider. Check the wiki for more information on setting
   # up on your models and hooks.
   # config.omniauth :github, 'APP_ID', 'APP_SECRET', scope: 'user,public_repo'
+#  config.omniauth :facebook, "1520032808310785", "507813b1f98b4c9bfc0686360ba10349", scope: "public_profile", info_fields: "id,name,link"
+  config.omniauth :google_oauth2,
+    "122867097489-nrcasn8t9ju90hlaufsom5kh2oo9k8qn.apps.googleusercontent.com",
```
Comment (dieunb) [inline config/initializers/devise.rb]: should move 
```
"122867097489-nrcasn8t9ju90hlaufsom5kh2oo9k8qn.apps.googleusercontent.com",
+    "55TFDGTTemyvu2qqqOaMu8Pd"
```
to environment variable

File: app/controllers/omniauth_callbacks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def all
```
Comment (dieunb) [inline app/controllers/omniauth_callbacks_controller.rb]: bad naming, should change it.

**Comment by MuhammadTamzid** [comment]
Ok

File: app/models/user.rb [hunk 7]
```diff
@@ -1,10 +1,43 @@
...
+        user.provider = auth.provider
+        user.uid = auth.uid
+        user.name = auth.info.name
+      end
+    end
+
+    def new_with_session params, session
+      if session["devise.user_attributes"]
+        new(session["devise.user_attributes"], without_protection: true) do |user|
+        user.attributes = params
```
Comment (dieunb) [inline app/models/user.rb]: indent

File: app/models/user.rb [hunk 8]
```diff
@@ -1,10 +1,43 @@
...
+        user.valid?
```
Comment (dieunb) [inline app/models/user.rb]: indent

File: config/heroku_env.rb [hunk 9]
```diff
@@ -0,0 +1,2 @@
+ENV["GOOGLE_ID"] = "122867097489-nrcasn8t9ju90hlaufsom5kh2oo9k8qn.apps.googleusercontent.com"
```
Comment (dieunb) [inline config/heroku_env.rb]: pls, read it https://help.ubuntu.com/community/EnvironmentVariables

Comment (sujoydatta) [inline config/heroku_env.rb]: Thank you

File: app/controllers/omniauth_callbacks_controller.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def omniauth_request
```
Comment (dieunb) [inline app/controllers/omniauth_callbacks_controller.rb]: Do you think about this  name: `identifier`?

Comment (sujoydatta) [inline app/controllers/omniauth_callbacks_controller.rb]: will change it in next pull

**Pull request merged**